src/main/java/org/javarosa/xpath/expr/XPathSortFunc.java

```diff
@@ -0,0 +1,146 @@
+package org.javarosa.xpath.expr;
+
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.xpath.XPathArityException;
+import org.javarosa.xpath.XPathTypeMismatchException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.crypto.dsig.spec.XPathType;
+
+/**
+ * Created by amstone326 on 6/28/17.
+ */
+
+public class XPathSortFunc extends XPathFuncExpr {
+
+    public static final String NAME = "sort";
+
+    // since we accept 1-3 arguments
+    private static final int EXPECTED_ARG_COUNT = -1;
+
+    public XPathSortFunc() {
+        name = NAME;
+        expectedArgCount = EXPECTED_ARG_COUNT;
+    }
+
+    public XPathSortFunc(XPathExpression[] args) throws XPathSyntaxException {
+        super(NAME, args, EXPECTED_ARG_COUNT, true);
+    }
+
+    @Override
+    protected void validateArgCount() throws XPathSyntaxException {
+        if (args.length < 1 || args.length > 3) {
+            throw new XPathArityException(name, "between 1 and 3 arguments", args.length);
+        }
+    }
+
+    @Override
+    protected Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        List<String> sortedList;
+        if (evaluatedArgs.length == 1) {
+            sortedList = sortSingleList(FunctionUtils.toString(evaluatedArgs[0]), true);
+        } else if (evaluatedArgs.length == 2) {
+            Object lastArg = evaluatedArgs[1];
+            if (lastArg instanceof Boolean) {
```

Comment:
Hm, fair point on the dynamic dispatching, but I don't agree that it's preferable to make people include the original list twice in a row if they want to sort a single list; I think that's the sort of thing that seems fine to us, but would mean that the spec for the function would make it very unclear to average users how they would ever use it to sort a single list. I would rather solve this by making it so that the directional argument is always required. Is that an OK solution?